src/app/fdctl/keys.c

```diff
@@ -46,72 +47,77 @@ keys_cmd_args( int *    pargc,
                  *pargv[0] ));
 }
 
-void
+void FD_FN_SENSITIVE
 generate_keypair( char const *     keyfile,
                   config_t * const config,
                   int              use_grnd_random ) {
-  uchar keys[ 64 ];
-
-  uint flags = use_grnd_random ? GRND_RANDOM : 0;
+  uint flags = use_grnd_random ? GRND_RANDOM : 0U;
 
+  uchar keypair[ 64 ];
   long bytes_produced = 0L;
-  while( FD_LIKELY( bytes_produced<32 ) ) {
-    long n = getrandom( keys+bytes_produced, (ulong)(32-bytes_produced), flags );
+  while( FD_LIKELY( bytes_produced<32L ) ) {
+    long n = getrandom( keypair+bytes_produced, (ulong)(32-bytes_produced), flags );
     if( FD_UNLIKELY( -1==n ) ) FD_LOG_ERR(( "could not create keypair, getrandom() failed (%i-%s)", errno, fd_io_strerror( errno ) ));
     bytes_produced += n;
   }
 
-  fd_sha512_t _sha[1];
+  fd_sha512_t _sha[ 1 ];
   fd_sha512_t * sha = fd_sha512_join( fd_sha512_new( _sha ) );
   if( FD_UNLIKELY( !sha ) ) FD_LOG_ERR(( "could not create keypair, fd_sha512 join failed" ));
-  fd_ed25519_public_from_private( keys+32, keys, sha );
+  fd_ed25519_public_from_private( keypair+32UL, keypair, sha );
 
-  /* switch to non-root uid/gid for file creation. permissions checks still done as root. */
+  /* Switch to non-root uid/gid for file creation.  Permissions checks
+     are still done as root. */
   gid_t gid = getgid();
   uid_t uid = getuid();
-  if( FD_LIKELY( gid == 0 && setegid( config->gid ) ) )
-    FD_LOG_ERR(( "setegid() failed (%i-%s)", errno, fd_io_strerror( errno ) ));
-  if( FD_LIKELY( uid == 0 && seteuid( config->uid ) ) )
-    FD_LOG_ERR(( "seteuid() failed (%i-%s)", errno, fd_io_strerror( errno ) ));
-
-  mode_t previous = umask( S_IRWXO | S_IRWXG | S_IXUSR );
-
-  FILE * fp = fopen( keyfile, "wx" );
-  if( FD_UNLIKELY( !fp ) ) {
-    if( FD_LIKELY( errno == EEXIST ) )
-      FD_LOG_ERR(( "could not create keypair as the keyfile `%s` already exists", keyfile ));
-    else
-      FD_LOG_ERR(( "could not create keypair, fopen(%s) failed (%i-%s)", keyfile, errno, fd_io_strerror( errno ) ));
+  if( FD_LIKELY( !gid && setegid( config->gid ) ) ) FD_LOG_ERR(( "setegid() failed (%i-%s)", errno, fd_io_strerror( errno ) ));
+  if( FD_LIKELY( !uid && seteuid( config->uid ) ) ) FD_LOG_ERR(( "seteuid() failed (%i-%s)", errno, fd_io_strerror( errno ) ));
+
+  /* find last `/` in keyfile and zero it */
+  char keyfile_copy[ PATH_MAX ];
+  fd_memcpy( keyfile_copy, keyfile, sizeof( keyfile_copy ) );
```

Comment:
I think `strncpy` be better so that it doesn't overrun `keyfile`, though that's not possible with current uses, so I'll stamp. I think really what this is suggesting is that the type of `keyfile` should be `char const keyfile[ static PATH_MAX ]` or something. I think just using strncpy is easier.